kernel/src/listed_log_files.rs

```diff
@@ -41,31 +41,84 @@ pub(crate) struct ListedLogFiles {
+/// take precedence.
 ///
-/// Note: this calls [`StorageHandler::list_from`] to get the list of log files.
+/// [`ParsedLogPath`] may be a commit or a checkpoint. If `start_version` is not specified, the
+/// files will begin from version number 0. If `end_version` is not specified, files up to the most
+/// recent version will be included.
+///
+/// Note: this may call [`StorageHandler::list_from`] to get the list of log files unless the
+/// provided log_tail covers the entire requested range.
+///
+/// TODO: consider renaming? `maybe_list_log_files`? `list_log_files_with_tail`?
 fn list_log_files(
     storage: &dyn StorageHandler,
     log_root: &Url,
+    log_tail: Vec<ParsedLogPath>,
     start_version: impl Into<Option<Version>>,
     end_version: impl Into<Option<Version>>,
 ) -> DeltaResult<impl Iterator<Item = DeltaResult<ParsedLogPath>>> {
+    #[cfg(debug_assertions)]
+    {
+        // log_tail is only commits
+        for entry in &log_tail {
+            debug_assert!(
+                matches!(entry.file_type, LogPathFileType::Commit),
+                "log_tail should only contain commits, found {:?}",
+                entry.file_type
+            );
+        }
+        // log_tail has no gaps nor duplicates
+        if !log_tail.is_empty() {
+            for i in 1..log_tail.len() {
+                debug_assert_eq!(
+                    log_tail[i].version,
+                    log_tail[i - 1].version + 1,
+                    "log_tail should be contiguous without gaps or duplicates"
+                );
+            }
+        }
+    }
+
+    // calculate listing bounds
     let start_version = start_version.into().unwrap_or(0);
     let end_version = end_version.into();
-    let version_prefix = format!("{start_version:020}");
-    let start_from = log_root.join(&version_prefix)?;
+    // start_from is log path to start listing from: the log root with zero-padded start version
+    let start_from = log_root.join(&format!("{start_version:020}"))?;
+    let list_end_version = match (log_tail.first(), end_version) {
+        // stop before log_tail or at requested end, whichever comes first
+        (Some(tail), Some(end)) => Some(end.min(tail.version.saturating_sub(1))),
+        // no end: stop just before log_tail starts
+        (Some(tail), None) => Some(tail.version.saturating_sub(1)),
+        // no tail: just respect requested end
+        (None, end) => end,
+    };
 
-    Ok(storage
+    // since engine APIs don't limit listing, we just list from start_version and filter
+    let listed_files = storage
         .list_from(&start_from)?
         .map(|meta| ParsedLogPath::try_from(meta?))
-        // TODO this filters out .crc files etc which start with "." - how do we want to use these kind of files?
+        // TODO this filters out .crc files etc which start with "." - how do we want to use these
+        // kind of files?
         .filter_map_ok(identity)
```

Comment:
Since we have the `LogPathFileType::Crc` variant now, I think this comment might be outdated?